### PR TITLE
Add helpers from darvaza/shared/x

### PR DESCRIPTION
**Pull request by amery**
* ListCopy/ListCopyFn
* MapListCopy/MapListCopyFn
* MapValueOr

to drop `darvaza.org/darvaza/shared/x`

## Summary by CodeRabbit

- **New Features**
	- Added functions for copying lists (`ListCopy`, `ListCopyFn`) and maps (`MapListCopy`, `MapListCopyFn`), enhancing data manipulation capabilities.
	- Introduced `MapValue` for retrieving values from maps with a default return option.

- **Documentation**
	- Updated the README.md to include new exported functions, improving clarity on library capabilities.

### CODE REVIEW

File: lists.go [hunk 1]
```diff
@@ -96,3 +96,26 @@ func listIterBackwardStep(ref *list.Element) (e *list.Element, prev *list.Elemen
...
+
+	out := list.New()
+	ListForEach(src, func(v0 T) bool {
+		if v1, ok := fn(v0); ok {
+			out.PushBack(v1)
+		}
+		return false
+	})
+	return out
+}
```
[the remark this answers is not in the document]
Comment (amery) [reply lists.go]: improvement added. thank you!

File: maps.go [hunk 2]
```diff
@@ -29,6 +29,15 @@ func SortedKeys[K Ordered, T any](m map[K]T) []K {
 	return keys
 }
 
+// MapValueOr returns a value of an entry or a default if
+// not found
+func MapValueOr[K comparable, V any](m map[K]V, key K, def V) (V, bool) {
```
Comment (amery) [inline maps.go:34]: I wonder if this should be just `MapValue()` instead

Comment (karasz) [reply maps.go]: I would prefer a `MapValue` over `MapValueOr`

Comment (amery) [reply maps.go]: > I would prefer a `MapValue` over `MapValueOr`

renamed

File: maps.go [hunk 3]
```diff
@@ -145,6 +154,23 @@ func MapListAppendUniqueFn[K comparable, T any](m map[K]*list.List, key K, v T,
...
+// MapListCopy duplicates a map containing a list.List
+func MapListCopy[T comparable](src map[T]*list.List) map[T]*list.List {
+	fn := func(v any) (any, bool) { return v, true }
+	return MapListCopyFn(src, fn)
+}
+
+// MapListCopyFn duplicates a map containing a list.List but
+// allows the element's values to be cloned by a helper function
+func MapListCopyFn[K comparable, V any](src map[K]*list.List,
+	fn func(v V) (V, bool)) map[K]*list.List {
```
Comment (karasz) [inline maps.go:166]: I think this would benefit from a 
```go
if src== nil || len(src)==0 {
return make(map[K]*list.List), false
}
```

**Review by karasz — CHANGES_REQUESTED** [review]

File: maps.go (hunk 3 above)
Comment (amery) [reply maps.go:166]: the second argument complicates significantly the usage when "cloning" objects

Comment (amery) [reply maps.go:166]: @karasz see https://github.com/darvaza-proxy/darvaza/commit/5ec539d06316f7de64446a69fe6247997a29eba0

Comment (karasz) [reply maps.go:166]: > @karasz see [darvaza-proxy/darvaza@5ec539d](https://github.com/darvaza-proxy/darvaza/commit/5ec539d06316f7de64446a69fe6247997a29eba0)

brain typo. Please read as
```go
return make(map[K]*list.List)
```

Comment (amery) [reply maps.go:166]: @karasz `golangci-lint` didn't like the idea
```
maps.go:168:5: S1009: should omit nil check; len() for map[K]*container/list.List is defined as zero (gosimple)
        if src == nil || len(src) == 0 {
           ^
exit status 1
```

```diff
@@ -164,6 +164,11 @@ func MapListCopy[T comparable](src map[T]*list.List) map[T]*list.List {
 // allows the element's values to be cloned by a helper function
 func MapListCopyFn[K comparable, V any](src map[K]*list.List,
        fn func(v V) (V, bool)) map[K]*list.List {
+       //
+       if src == nil || len(src) == 0 {
+               return make(map[K]*list.List)
+       }
+
        out := make(map[K]*list.List, len(src))
        for k, l := range src {
                out[k] = ListCopyFn(l, fn)
```

Comment (amery) [reply maps.go:166]: @karasz what about this instead?

```diff
@@ -164,9 +164,15 @@ func MapListCopy[T comparable](src map[T]*list.List) map[T]*list.List {
 // allows the element's values to be cloned by a helper function
 func MapListCopyFn[K comparable, V any](src map[K]*list.List,
        fn func(v V) (V, bool)) map[K]*list.List {
-       out := make(map[K]*list.List, len(src))
-       for k, l := range src {
-               out[k] = ListCopyFn(l, fn)
+       //
+       l := len(src)
+       if l == 0 {
+               return make(map[K]*list.List)
+       }
+
+       out := make(map[K]*list.List, l)
+       for k, lp := range src {
+               out[k] = ListCopyFn(lp, fn)
        }
        return out
 }
 ```

Comment (karasz) [reply maps.go:166]: Not worth it. Just drop it as it is, after more consideration feels like micro optimization.

**Review by karasz — APPROVED** [review]
LGTM

**Comment by amery** [comment]
`v0.14.9` tagged

**Pull request merged**